### ISSUE TITLE
More general type for "while true do ... done".

### ISSUE THREAD

**Issue by rlepigre** [ocaml/ocaml#11805]
I often find myself writing loops that are only exited via an exception, for example when reading a file:
```ocaml
let input_lines : in_channel -> string list = fun ic ->
  let lines = ref [] in
  try while true do
    lines := input_line ic :: !lines
  done; assert false with End_of_file ->
  List.rev !lines
```
However, as you can see in the example above, one needs to write something like `while true do ... done; assert false`, even though the loop can never be exited normally.

Would it make sense to have a special case for `while true do ... done` in type-inference, so that it is given a weakly-polymorphic type?

**Comment by nojb** [comment]
Seems an overly special use-case to justify such a large change to the language. By the way, here's an alternative way to write the loop that does away with the `assert false`:
```ocaml
let lines = ref [] in
while
  match input_line ic with
  | line -> lines := line :: !lines; true
  | exception End_of_file -> false
do () done;
List.rev !lines
```

**Comment by rlepigre** [comment]
Yeah, I see how one could argue that it is a weird special case. However, I would not have expected this to be a large change, since it only concerns the type system. Basically, what I would expect would be to change [that line](https://github.com/ocaml/ocaml/blob/74068a7fc30ddddbb3910abbbac8b044457f1653/typing/typecore.ml#L3382) conditionally on `scond` being the value `true`. Seems pretty localized to me.

Also, the code I gave is just an example of a pattern I've encountered plenty of times. I know the code can be rewritten to not have the `assert false`. In fact, the feature I'd ideally like is a proper `break` (related to what I said in https://github.com/ocaml/ocaml/issues/11740).

**Comment by johnwhitington** [comment]
I proposed this excellent idea before:

https://github.com/ocaml/ocaml/issues/6833

Others did not agree!

**Comment by nojb** [comment]
> Seems pretty localized to me.

Right; it _is_ true that the patch is trivial. What I had in mind by "large" is that extending the `assert false` hack to other constructs feels like a "big" decision. But I'm not strongly opposed to this change either; let's wait for other devs to chime in, perhaps opinions have changed since #6833.

**Comment by Octachron** [comment]
I am not convinced by the idea of adding a special case to the type system to support an alternative way to write:
```ocaml
let rec neverendingly f = f (); neverendingly f
let input_lines ic =
  let lines = ref [] in
  try neverendingly (fun () ->
    lines := input_line ic :: !lines
   ) with End_of_file ->
  List.rev !lines
```

**Comment by gasche** [comment]
Personally I think that the suggestion has merit. I'm fine with `assert false`, and I think `while true do` is also natural. (I'm not averse to special-casing subsets of a syntactic construction, the downsides compared to introducing new syntax for these subsets are not that clear.). This pattern seems to occur often enough at non-unit types that `done; assert false` is found fairly often in OCaml code in the wild:

https://github.com/search?p=1&q=%22done%3B+assert+false%22+extension%3Aml+language%3AOCaml&type=Code

(100 pages of result, that seem to come from diverse programs doing various things)

Having a `forever : (unit -> unit) -> 'a` function in the standard library could also work, but I am not sure where it should go. `Fun.forever` ?

**Comment by rlepigre** [comment]
> I proposed this excellent idea before:

No idea how I did not find your existing issue: I did run a quick search. :)

**Comment by rlepigre** [comment]
> I am not convinced by the idea of adding a special case to the type system to support an alternative way to write:
> 
> ```ocaml
> let rec neverendingly f = f (); neverendingly f
> let input_lines ic =
>   let lines = ref [] in
>   try neverendingly (fun () ->
>     lines := input_line ic :: !lines
>    ) with End_of_file ->
>   List.rev !lines
> ```

To me, it's a bit like saying that `while` loops are useless, since you can always replace them by a recursive function. But sure, there are many alternative implementations for the example I gave.

**Comment by johnwhitington** [comment]
> Seems an overly special use-case to justify such a large change to the language. By the way, here's an alternative way to write the loop that does away with the `assert false`:
> 
> ```ocaml
> let lines = ref [] in
> while
>   match input_line ic with
>   | line -> lines := line :: !lines; true
>   | exception End_of_file -> false
> do () done;
> List.rev !lines
> ```

I find while loops with empty bodies and full conditions inexplicably harder to think about, in the same way that a schoolchild can understand `x = 1` but get confused by `1 = x`. This may be just me though.

**Comment by nojb** [comment]
Is it possible to converge? From the discussion I don't see any real opposition. @damiendoligez: as the dissenting voice in #6833, do you care to share your opinion on this proposal, seven years later?

**Comment by gasche** [comment]
My understanding is that @Octachron is also rather strongly opposed to the proposed feature -- I'm not sure why. (And also in vacation right now.)

**Comment by johnwhitington** [comment]
I don't find the "you should write it this way instead" objections very convincing at all. OCaml has imperative constructs, and they are useful. Introducing extra functions called "loop", for example, may be functional, but it is not natural. The alternative imperative formulations are worse too.

The other objections seem to be on ugliness (it's not ugly) or type-theoretic grounds (not really qualified to comment on those)

In my opinion, this proposal:

- is natural. An infinite while loop cannot complete execution except via an exception, so we shouldn't need to specify a type for it. Let f x = f x doesn't return either, so it gets 'a. Fine.
- helps beginners. We shouldn't have to explain an "assert false" trick or an "add a value of the right type which will never be reached after the loop" trick to a beginner. Tricks are bad.
- is simple and unobtrusive to implement, and entirely backward compatible
- Reduces the use of assert false. It is not adding a new wart to the language, it is replacing a big wart with a smaller one.

**Comment by Octachron** [comment]
I am not fond of the feature, I am not strongly opposed to it, at least when described as a way to avoid adding a new `loop` or `forever` keyword by making `while true` a 2-word keyword (a keydigram?) like `assert false`.

I am bit worried when I read the argument in favor of the feature however, since they tend to convince me that the feature is:

* slightly unnatural: it introduces a second place in the language where literal cannot be replaced by 
an equivalent constant
```ocaml
let debug = true
let () = while debug do ... done
``` 
is no longer equivalent to
```ocaml
let () = while true do ... done
```
Bonus question: what happens with modular macros (or MetaOCaml)? I expect that in both cases, we end up in the first situation (which is one more reason to explain the feature as a composite keyword). 
* may confuse beginners by adding a special rule to the type system that only matters in a specific use case. The type of `while ... do ... done` cannot be described anymore in a simple proposition since we have to either explain that `while true do ... done` is a different construction, or explain that there is a trick in the language to make the type of `while true do ... done` different. Contrarily, there are no special rules for the typing of `let rec f () = f ()`, it follows naturally from the rules for typing recursive functions.
* is a breaking change:
```ocaml
class c = object
  method loop () = while true do raise Exit done
end
```
I don't think this matters too much, but I would not be surprised if this does break some code in the wild.
* increase the complexity of the language  to avoid adding a single natural function to the standard library of a functional language.

I think that the feature is much more natural if described as introducing `while true` as a 2-word keyword. Moreover, I don't write enough `while` loop to feel qualified enough for my advice to matter on this subject, so I will gladly let people that write `while` loop decide.

**Comment by dbuenzli** [comment]
Personally I doubt the whole "it helps beginners", "it's natural" comments. 

A special case is being added which adds cognitive burden to users of the language and I'm pretty sure @Octachron's example will confuse everyone beginners or not: referential transparency is good.

**Comment by yallop** [comment]
> it introduces a second place in the language where literal cannot be replaced by an equivalent constant

There are quite a few of these besides `assert false` already.  For example,

1. this is permitted:
   ```ocaml
   let rec x = ref (lazy (ignore x))
   ```
   but this is rejected:
   ```ocaml
   let reference = ref
   let rec x = reference (lazy (ignore x))
   ```
   
2. this evaluates to `true`
   ```ocaml
   (||) true (assert false)  
   ```
   but this raises an exception
   ```ocaml
   let either = (||)  in
   either true (assert false)
   ```

3. this is permitted:
   ```ocaml
   module F () = Set
   module G = F(struct end)
   ```
   but this is rejected:
   ```ocaml
   module F () = Set
   module H = struct end
   module G = F(H)
   ```

None of this is a strong argument in favour of the feature, of course, but I think it does show that the loss of referential transparency is not particularly unusual.
   
> Bonus question: what happens with modular macros (or MetaOCaml)? I expect that in both cases, we end up in the first situation

Indeed, I think it's quite likely that we'd perform the syntactic check at macro-definition time, and so give `<<while $(<<true>) ... >>` the less-general type (much as MetaOCaml already rejects `.< let rec x = .~(.<lazy (ignore x)>.) in x >.`).

On the feature itself: I'm still in favour, but my support has cooled slightly since #6833.  Both this PR and #6833 use `input_lines` as the motivating example, but nowadays that function can be written more naturally with a combination of `match`/`exception` and TRMC, as @xavierleroy has shown in #11843.

**Comment by damiendoligez** [comment]
I'm not strongly opposed to this feature but:
1. It touches the type-checker, which is the most complex part of the system.
2. It introduces a special case (best explained as a two-word keyword, as @Octachron said).
3. It seems that nowadays whenever I start writing `while` I later end up replacing it with a tail-recursive function.

Note @rlepigre 's remark:
> In fact, the feature I'd ideally like is a proper `break`

When you use a tail-recursive function instead of a loop, you get your `break` and much more (natural expression of control flows that cannot be expressed with a simple loop) and you usually eliminate a few references at the same time.

**Comment by nojb** [comment]
It is more or less clear that there isn't enough positive support for this change among the maintainers that have chimed in. So I think we will have to table this discussion and revisit it in the future.

**Comment by nojb** [comment]
This issue was discussed in today's developer meeting and, against all expectations, there was a consensus in favour of this change. All we need now is a candidate to prepare a PR.

**Comment by johnwhitington** [comment]
> This issue was discussed in today's developer meeting and, against all expectations, there was a consensus in favour of this change. All we need now is a candidate to prepare a PR.

Jeremy Yallop's patch from the linked bug could be used as a basis:

https://gist.githubusercontent.com/vicuna/12a91469e1a704ec05d6de30d97ccd09/raw/2198eee98a2bcf9819473b79a76765b9c03fca28/while_true.patch

**Comment by yallop** [comment]
I'd be happy to update the patch (if necessary) and turn it into a PR.

**Comment by nojb** [comment]
> I'd be happy to update the patch (if necessary) and turn it into a PR.

That would be great, thanks!

**Comment by gasche** [comment]
The feature is now merged upstream and should be part of OCaml 5.2.
